README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

This is a multi axis scrollable paginated data table, that allows you to scroll on both the vertical and horizontal axis, and also gives you the ability to style columns
and rows cells. please look at the demo blow.

## Features

### Demo

![alt text](https://i.postimg.cc/8P8jpjxc/Screenshot-4.png)

This widget serves the same purpose as a DataTable.

## Getting started

Simply add into your dependencies the following line.

```dart
dependencies:
  flutter_easy_table: ^0.0.1
```

## Usage

```dart
EasyPaginatedTable(
        height: 300,
        width: 600,
        rowTail: true,
        rowsPerPage: 3,
        columnStyle: ColumnStyle(
          columnLabel: 'name',
          columnStyle: const TextStyle(
            color: Colors.blue,
            fontWeight: FontWeight.bold,
          ),
          rowCellLabel: 'Taha',
          rowCellStyle: const TextStyle(
            color: Colors.green,
            fontWeight: FontWeight.bold,
          ),
        ),
        onEdit: (index) {},
        onDelete: (index) {},
        columns: const ['id', 'name', 'age'],
        rows: const [
          {'id': '1', 'name': 'Taha', 'age': '27'},
          {'id': '2', 'name': 'wahab', 'age': '33'},
          {'id': '3', 'name': 'ahmed', 'age': '30'},
          {'id': '4', 'name': 'Taha', 'age': '27'},
          {'id': '5', 'name': 'wahab', 'age': '33'},
          {'id': '6', 'name': 'ahmed', 'age': '30'},
        ],
      )
```

## Additional information

GitHub Repo: <https://github.com/Tahateber95/flutter_easy_table>


CHANGELOG.md
## 0.0.1

* A multi axis scrollable paginated data table, that allows you to scroll on both the vertical and horizontal axis with custom styling ability.


test/easy_table_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:easy_table/easy_table.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/src/easy_paginated_table.dart
import 'package:flutter/material.dart';
import 'package:iconsax/iconsax.dart';

class EasyPaginatedTable extends StatelessWidget {
  EasyPaginatedTable(
      {Key? key,
      required this.rows,
      required this.columns,
      this.onEdit,
      this.onDelete,
      this.rowTail = false,
      required this.width,
      required this.height,
      this.columnStyle,
      this.rowsPerPage = 10,
      this.columnSpacing = 100,
      this.horizontalMargin = 5})
      : super(key: key);
  final List<Map<String, String?>> rows;
  final List<String> columns;
  final ColumnStyle? columnStyle;
  final bool rowTail;
  final void Function(int)? onEdit;
  final void Function(int)? onDelete;
  final double width, height;
  final int rowsPerPage;
  final double columnSpacing;
  final double horizontalMargin;
  @override
  Widget build(BuildContext context) {
    final DataTableSource data = MyData(
        rows: rows,
        rowTail: rowTail,
        onEdit: onEdit,
        onDelete: onDelete,
        columns: columns,
        columnStyle: columnStyle);

    final ScrollController controller = ScrollController();

    return SizedBox(
      height: height,
      width: width,
      child: Scrollbar(
        controller: controller,
        child: SingleChildScrollView(
          controller: controller,
          scrollDirection: Axis.horizontal,
          child: SizedBox(
            width: MediaQuery.of(context).size.width,
            height: height,
            child: ListView(
              shrinkWrap: true,
              primary: false,
              children: [
                PaginatedDataTable(
                  source: data,
                  columns: [
                    ...List.generate(
                        columns.length,
                        (index) => DataColumn(
                                label: Padding(
                              padding: const EdgeInsets.all(8.0),
                              child: Text(columns[index]),
                            ))),
                    rowTail
                        ? const DataColumn(
                            label: Padding(
                            padding: EdgeInsets.all(8.0),
                            child: Text(''),
                          ))
                        : DataColumn(label: Text(''))
                  ],
                  columnSpacing: 100,
                  horizontalMargin: 5,
                  rowsPerPage: rowsPerPage,
                  showCheckboxColumn: false,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class MyData extends DataTableSource {
  MyData({
    Key? key,
    required this.rows,
    required this.columns,
    this.columnStyle,
    this.rowTail = false,
    required this.onEdit,
    required this.onDelete,
  });

  final List<Map<String, String?>> rows;
  final List<String> columns;
  final bool rowTail;
  final ColumnStyle? columnStyle;
  final void Function(int)? onEdit;
  final void Function(int)? onDelete;

  @override
  bool get isRowCountApproximate => false;
  @override
  int get rowCount => rows.length;
  @override
  int get selectedRowCount => 0;
  @override
  DataRow getRow(int index) {
    return DataRow(cells: [
      ...columns.map((column) {
        return DataCell(Padding(
            padding: const EdgeInsets.all(8.0),
            child: columnStyle != null
                ? (column == columnStyle!.columnLabel &&
                        rows[index][column] == columnStyle!.rowCellLabel)
                    ? Text(
                        rows[index][column] ?? "",
                        style: columnStyle!.rowCellStyle,
                      )
                    : (column == columnStyle!.columnLabel)
                        ? Text(
                            rows[index][column] ?? "",
                            style: columnStyle!.columnStyle,
                          )
                        : (column == columnStyle!.columnLabel &&
                                columnStyle!.rowCellLabel == null)
                            ? Text(
                                rows[index][column] ?? "",
                                style: columnStyle!.columnStyle,
                              )
                            : Text(
                                rows[index][column] ?? "",
                                style: TextStyle(),
                              )
                : Text(
                    rows[index][column] ?? "",
                    style: TextStyle(),
                  )));
      }).toList(),
      rowTail
          ? DataCell(Padding(
              padding: const EdgeInsets.all(8.0),
              child: Row(
                children: [
                  IconButton(
                    onPressed: () {
                      onEdit!(index);
                    },
                    icon: const Icon(Iconsax.edit),
                  ),
                  const SizedBox(width: 5.0),
                  IconButton(
                    onPressed: () {
                      onDelete!(index);
                    },
                    icon: const Icon(Iconsax.note_remove, color: Colors.red),
                  ),
                ],
              ),
            ))
          : const DataCell(Text(''))
    ]);
  }
}

class ColumnStyle {
  ColumnStyle(
      {required this.columnLabel,
      required this.columnStyle,
      this.rowCellLabel,
      this.rowCellStyle});

  /// The specfic column's name that needs to be styled.
  final String columnLabel;

  /// The specfic rows's name that needs to be styled.
  final String? rowCellLabel;
  final TextStyle? rowCellStyle;
  final TextStyle columnStyle;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_easy_table/flutter_easy_table.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Easy Table'),
      ),
      body: EasyPaginatedTable(
        height: 300,
        width: 600,
        rowTail: true,
        rowsPerPage: 3,
        columnStyle: ColumnStyle(
          columnLabel: 'name',
          columnStyle: const TextStyle(
            color: Colors.blue,
            fontWeight: FontWeight.bold,
          ),
          rowCellLabel: 'Taha',
          rowCellStyle: const TextStyle(
            color: Colors.green,
            fontWeight: FontWeight.bold,
          ),
        ),
        onEdit: (inded) {},
        onDelete: (index) {},
        columns: const ['id', 'name', 'age'],
        rows: const [
          {'id': '1', 'name': 'Taha', 'age': '27'},
          {'id': '2', 'name': 'wahab', 'age': '33'},
          {'id': '3', 'name': 'ahmed', 'age': '30'},
          {'id': '4', 'name': 'Taha', 'age': '27'},
          {'id': '5', 'name': 'wahab', 'age': '33'},
          {'id': '6', 'name': 'ahmed', 'age': '30'},
        ],
      ),
    );
  }
}
